### DISCUSSION TITLE
Wrong favicons for ports and subdomains

### DISCUSSION THREAD

**Discussion by rakuri255** [dani-garcia/vaultwarden#3756] [Q&A]
### Wrong favicons for ports and subdomains
If you have applications with own favicons, which are in host:port or in subdomain.domain, only host:80 and domain will be used. So you always have the wrong favicons. But it works when you create a new entry for a subdomain.

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.29.1
* Web-vault version: v2023.5.0
* OS/Arch: linux/x86_64
* Running within Docker: true (Base: Debian)
* Environment settings overridden: false
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Browser/Server Time Check: true
* Server/NTP Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.41.2
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** 

```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": false,
  "_enable_smtp": true,
  "_enable_yubico": false,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "_smtp_img_src": "cid:",
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_session_lifetime": 20,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "***************",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://*******************************",
  "domain_origin": "*****://*******************************",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 3 * * * *",
  "emergency_request_timeout_schedule": "0 7 * * * *",
  "enable_db_wal": true,
  "event_cleanup_schedule": "0 10 0 * * *",
  "events_days_retain": null,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": false,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "Vaultwarden",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "Info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "org_events_enabled": false,
  "org_groups_enabled": false,
  "password_hints_allowed": true,
  "password_iterations": 100000,
  "push_enabled": false,
  "push_installation_id": "***",
  "push_installation_key": "***",
  "push_relay_uri": "https://push.bitwarden.com",
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sendmail_command": null,
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": false,
  "signups_domains_whitelist": "",
  "signups_verify": false,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_embed_images": true,
  "smtp_explicit_tls": null,
  "smtp_from": "************************",
  "smtp_from_name": "Vaultwarden",
  "smtp_host": "**************",
  "smtp_password": "***",
  "smtp_port": 587,
  "smtp_security": "starttls",
  "smtp_ssl": null,
  "smtp_timeout": 15,
  "smtp_username": "*********",
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_sendmail": false,
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": false,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```
</details>

* vaultwarden version: 1.29.1

* Install method: Docker image

* Clients used: web vault, desktop, Android, iOS

### Steps to reproduce
Create some webpages with different favicons

Web 1: http://192.168.0.218:80 -> Favicon 1
Web 2: http://192.168.0.218:8888 -> Favicon 2
Web 3: http://192.168.0.218:9999 --> Favicon 3

Web 1: https://mainDomain.com -> Favicon 1
Web 2: https://subDomainOne.mainDomain.com -> Favicon 2
Web 3: https://subDomainTwo.mainDomain.com--> Favicon 3

Empty the server icon cache, after saving all entry to vaultwarden.

### Expected behaviour
Should have a different favicon for each host:port and for each subdomain

### Actual behaviour
Shows only the icon from the host

Web 1: http://192.168.0.218:80 -> Favicon 1
Web 2: http://192.168.0.218:8888 -> Favicon 1
Web 3: http://192.168.0.218:9999 --> Favicon 1

Web 1: https://mainDomain.com -> Favicon 1
Web 2: https://subDomainOne.mainDomain.com -> Favicon 1
Web 3: https://subDomainTwo.mainDomain.com--> Favicon 1

### Troubleshooting data
It works when you  create a new entry for a subdomain.
But it dont works when the icon cache is empty and need to redownload the favicon.

**Comment by BlackDex** [comment]
Ports do not work, those are not passed by the clients, so nothing we can do there.

If the subdomains do not provide a website, or need login or do not have any valid `<link>` tags for the icons it will try the main domain.

You could enable debug log_level which should provide some info onto why it gets skipped, just make sure you cleared the cache before doing the test, else it will skip it.

**Comment by rakuri255** [comment]
@BlackDex  Thanks for the info with the port! 
I just noticed that it also depends on the order of the saved url. 
So the subdomains works, but only if they are saved as first URL. 
If the first URL is a host:port and second URL is the subdomain.domain, only the first URL is tested for favicon. Which is ok if there is actually a favicon. 
But when there is no favicon, which is then saved with host.png.miss. Then the other saved URLs wont get tested.
